src/mpid/ch4/shm/xpmem/xpmem_control.h

```diff
@@ -149,4 +157,164 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_ctrl_send_lmt_req_cb(MPIDI_SHM_ctrl_hdr
     goto fn_exit;
 }
 
+#ifdef MPIDI_CH4_SHM_XPMEM_COOP_P2P
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_ctrl_send_lmt_coop_req_cb
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_ctrl_send_lmt_coop_req_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr,
+                                                                   int packet_type)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_req_t *slmt_req_hdr = &ctrl_hdr->xpmem_slmt_req;
+    MPIR_Request *rreq = NULL;
+    MPIR_Request *sreq = NULL;
+    MPIDI_av_entry_t *av = NULL;
+    MPIR_Comm *root_comm;
+    MPIR_Request *anysource_partner;
+    int recvtype_iscontig, counter;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_CTRL_SEND_LMT_COOP_REQ_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_CTRL_SEND_LMT_COOP_REQ_CB);
+
+    if (packet_type == MPIDI_SHM_XPMEM_SEND_LMT_CTS) {
+        /* Sender gets the CTS packet from receiver and need to perform copy */
+        sreq = (MPIR_Request *) slmt_req_hdr->sreq_ptr;
+        MPIDIG_REQUEST(sreq, rank) = slmt_req_hdr->src_rank;
+        MPIDIG_REQUEST(sreq, tag) = slmt_req_hdr->tag;
+        MPIDIG_REQUEST(sreq, context_id) = slmt_req_hdr->context_id;
+
+        /* Complete XPMEM receive */
+        mpi_errno =
+            MPIDI_XPMEM_handle_lmt_coop_recv(slmt_req_hdr->src_offset, slmt_req_hdr->data_sz,
+                                             slmt_req_hdr->rreq_ptr, slmt_req_hdr->src_lrank,
+                                             MPIDI_SHM_XPMEM_SEND_LMT_CTS, sreq);
+        if (MPI_SUCCESS != mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+    } else {
+        XPMEM_PT2PT_DBG_PRINT
+            ("send_lmt_req_cb: received src_offset 0x%lx, data_sz 0x%lx, sreq_ptr 0x%lx, "
+             "src_lrank %d, match info[src_rank %d, tag %d, context_id 0x%x]\n",
+             slmt_req_hdr->src_offset, slmt_req_hdr->data_sz, slmt_req_hdr->sreq_ptr,
+             slmt_req_hdr->src_lrank, slmt_req_hdr->src_rank, slmt_req_hdr->tag,
+             slmt_req_hdr->context_id);
+
+        /* Try to match a posted receive request.
+         * root_comm cannot be NULL if a posted receive request exists, because
+         * we increase its refcount at enqueue time. */
+        root_comm = MPIDIG_context_id_to_comm(slmt_req_hdr->context_id);
```

Comment:
Let's define some common routines for handling the queued request. It is crazy to copy/paste such long code...